DOCS-5242 Update OAuth connections custom flow guide (#1045)
clerk/clerk-docs: docs/custom-flows/oauth-connections.mdx

# OAuth connections

<Callout type="danger">
  This guide is for users who want to build a *custom* user interface using the Clerk API. To authenticate users with OAuth connections using a *prebuilt* UI, you should use Clerk's [Account Portal pages](/docs/account-portal/overview) or [prebuilt components](/docs/components/overview).
</Callout>

## Before you start

You must configure your application instance through the Clerk Dashboard for the social connection(s) that you want to use, as described at [the top of the OAuth guide](/docs/authentication/social-connections/oauth#configuration).

## Create the sign-up and sign-in flow

> When using OAuth, the sign-up and sign-in are equivalent. A successful OAuth flow consists of the following steps:

Reviewer: ```suggestion
When using OAuth, the sign-up and sign-in flows are equivalent. A successful OAuth flow consists of the following steps:
```

1. Start the OAuth flow by calling [`SignIn.authenticateWithRedirect(params)`](/docs/references/javascript/sign-in/authenticate-with#authenticate-with-redirect) or [`SignUp.authenticateWithRedirect(params)`](/docs/references/javascript/sign-up/authenticate-with#authenticate-with-redirect). Note that both of these methods require a `redirectUrl` param, which is the URL that the browser will be redirected to once the user authenticates with the OAuth provider.

> 2. Create a route at the URL that the `redirectUrl` param points to. It's recommended to name this route `/sso-callback`. This route should call the [`Clerk.handleRedirectCallback()`](/docs/references/javascript/clerk/handle-navigation#handle-redirect-callback) method or simply render the prebuilt [`<AuthenticateWithRedirectCallback/>`](/docs/components/control/authenticate-with-callback) component.

Reviewer: Why do we recommend naming it `/sso-callback`? Are there any technical reasons, or is this just the most common name for this kind of route?

Author: @royanger do you have an answer here?

Author: it's just a suggestion- not a hard recommendation. I updated the copy 👍

1. Start the OAuth flow by calling [`SignIn.authenticateWithRedirect(params)`](/docs/references/javascript/sign-in/authenticate-with#authenticate-with-redirect) or [`SignUp.authenticateWithRedirect(params)`](/docs/references/javascript/sign-up/authenticate-with#authenticate-with-redirect). Note that both of these methods require a `redirectUrl` param, which is the URL that the browser will be redirected to once the user authenticates with the OAuth provider.
2. Create a route at the URL that the `redirectUrl` param points to. The following example names this route `/sso-callback`. This route should call the [`Clerk.handleRedirectCallback()`](/docs/references/javascript/clerk/handle-navigation#handle-redirect-callback) method or simply render the prebuilt [`<AuthenticateWithRedirectCallback/>`](/docs/components/control/authenticate-with-callback) component.

The following example shows two files:

1. The sign-in page where the user can start the OAuth flow.
2. The SSO callback page where the OAuth flow is completed.

<Tabs items={["Next.js"]}>
  <Tab>
  <CodeBlockTabs options={["Sign in page", "SSO callback page"]}>

>     ```jsx filename="app/sign-in/[[...sign-in]]/page.tsx"

Reviewer 2: ```suggestion
    ```jsx filename="app/sign-in/page.tsx"
```
Optional catchall is not needed for this flow, especially with the strictly defined `app/sign-up/sso-callback/page.tsx` example.

      console.log('testing');
      // Handle the redirect flow by rendering the
      // prebuilt AuthenticateWithRedirectCallback component.
      // This is the final step in the custom OAuth flow.
      return <AuthenticateWithRedirectCallback />;
    }
    ```
  </CodeBlockTabs>
  </Tab>
</Tabs>

## OAuth for React Native applications

With Clerk, you can add OAuth flows in your [React Native or Expo applications](/docs/quickstarts/expo).

Clerk ensures that security critical nonces will be passed only to allowlisted URLs when the OAuth flow is complete in native browsers or webviews.

So for maximum security in your production instances, you need to allowlist your custom redirect URLs via the [Clerk Dashboard](https://dashboard.clerk.com/) or the [Clerk Backend API](/docs/references/backend/redirect-urls/create-redirect-url).

To allowlist a redirect URL via the Clerk Dashboard:
1. Navigate to the [Clerk Dashboard](https://dashboard.clerk.com/last-active?path=user-authentication/social-connections).
2. In the navigation sidebar, select **User & Authentication > Social Connections**.
3. Scroll to the **Allowlist for mobile OAuth redirect** section and add your redirect URLs.

> ## OAuth account transfer flows

Reviewer 2: This could all be part of the main OAuth custom flow example. A transfer flow is super common, and should be added to most.

There are some edge cases where customers don't want it, so maybe keeping it separate makes sense. If so, then I would move this up so its under the '[Create the sign-up and sign-in flow](https://clerk.com/docs/pr/1045/custom-flows/oauth-connections#create-the-sign-up-and-sign-in-flow)' and this would really be a part 2 for most customers.

I would change the paragraph to be more like (rough draft):
----
It is common for users with OAuth to use a sign-in button when they need to sign-up, or vice versa. For those cases the `useSignIn` and `useSignUp` hooks have a `transferable` status that you can use to gracefully switch flows programmatically as needed. Most customers will want to add this to their custom flow unless they want to strictly separate sign-ins and sign-ups.

In some cases, a user may be trying to sign in with an OAuth account, but they don't have an account in your application yet. Or a user may be trying to sign up with an OAuth account, but they already have an account in your application. In these cases, Clerk provides "account transfers" which allow you to forward the user's information to the appropriate flow.

The following example shows how to handle these cases in your sign-in flow. The same logic can be applied to the sign-up flow; change the `signIn.AuthenticateWithRedirect()` to `signUp.authenticateWithRedirect()`.

<Tabs items={["Next.js"]}>
  <Tab>

>   ```jsx filename="app/sign-in/[[...sign-in]]/page.tsx"

Reviewer 2: ```suggestion
  ```jsx filename="app/sign-in/page.tsx"
```
Optional catchall is not needed for this flow, especially with the strictly defined `app/sign-up/sso-callback/page.tsx` example.

      const userExistsButNeedsToSignIn =
        signUp.verifications.externalAccount.status === 'transferable' &&
        signUp.verifications.externalAccount.error?.code ===
          'external_account_exists';

      if (userExistsButNeedsToSignIn) {
        const res = await signIn.create({ transfer: true });

        if (res.status === 'complete') {
          setActive({
            session: res.createdSessionId,
          });
        }
      }

      // If the user has an OAuth account but does not yet
      // have an account in your app, you can create an account
      // for them using the OAuth information.
      const userNeedsToBeCreated =
        signIn.firstFactorVerification.status === 'transferable';

      if (userNeedsToBeCreated) {
        const res = await signUp.create({
          transfer: true,
        });

        if (res.status === 'complete') {
          setActive({
            session: res.createdSessionId,
          });
        }
      } else {
        // If the user has an account in your application
        // and has an OAuth account connected to it, you can sign them in.
        signInWith(strategy);
      }
    }

>     const signInWith = (strategy: OAuthStrategy) => {

Reviewer 2: nit: Would love to see this moved up so the function is defined before its called.

> So for maximum security in your production instances, you need to allowlist your custom redirect URLs via the [Clerk Dashboard](https://dashboard.clerk.com/) or the [Clerk Backend API](https://clerk.com/docs/reference/backend-api/tag/Redirect-URLs).

Reviewer: I know you didn't change this sentence, but the link here doesn't seem to take you to any information about allowlisting a custom redirect URL, but maybe I missed it somewhere on the page.

Author: yes, its the create method. 
https://clerk.com/docs/reference/backend-api/tag/Redirect-URLs#operation/CreateRedirectURL

Author: I updated the copy to link to the backend sdk method instead 👍

To allowlist a redirect URL via the Clerk Dashboard:
1. Navigate to the [Clerk Dashboard](https://dashboard.clerk.com/last-active?path=user-authentication/social-connections).
2. In the navigation sidebar, select **User & Authentication > Social Connections**.

> 3. Select the **Setting** tab and add your redirect URL to the **Redirect URLs** list.

Author: updated the instructions!

## OAuth account transfer flows

> In some cases, a user may be trying to sign in with an OAuth account, but they don't have an account in your application yet. Or a user may be trying to sign up with an OAuth account, but they already have an account in your application. In these cases, Clerk provides "account transfers" which allow you to forward the user's information to the appropriate flow.

Reviewer: ```suggestion
In some cases, a user may try to sign in with OAuth without having an account with your app, or sign up with OAuth despite already having an account with your app. In these cases, Clerk provides "account transfers", which allow you to forward the user's information to the appropriate flow.
```
Simplified

Author: I can remove the quotes from account transfers, but I think the language I used is more clear.
The semantics you suggested are kind of confusing upon reading. I think it's better to break up the sentences and make the two use cases very clearly defined.

Reviewer: Ignore my suggestion then

> The following example shows how to handle these cases in your sign-in flow. The same logic can be applied to the sign-up flow; simply change the `signIn.AuthenticateWithRedirect()` to `signUp.authenticateWithRedirect()`.

Reviewer: ```suggestion
The following example shows how to handle these cases in your sign-in flow. The same logic can be applied to the sign-up flow; change the `signIn.AuthenticateWithRedirect()` to `signUp.authenticateWithRedirect()`.
```

<Tabs items={["Next.js"]}>
  <Tab>
  ```jsx filename="app/sign-in/[[...sign-in]]/page.tsx"
  'use client';

  import * as React from 'react';
  import { OAuthStrategy } from '@clerk/types';

>   import { SignInButton, useSignIn, useSignUp } from '@clerk/nextjs';

Reviewer: ```suggestion
  import { useSignIn, useSignUp } from '@clerk/nextjs';
```
Never used
